### PR TITLE
[SPARK-13091][SQL] Rewrite/Propagate constraints for Aliases

### PR DESCRIPTION
**Pull request by sameeragarwal**
This PR adds support for rewriting constraints if there are aliases in the query plan. For e.g., if there is a query of form `SELECT a, a AS b`, any constraints on `a` now also apply to `b`.

JIRA: https://issues.apache.org/jira/browse/SPARK-13091

cc @marmbrus

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala [hunk 1]
```diff
@@ -69,6 +72,20 @@ class ConstraintPropagationSuite extends SparkFunSuite {
         IsNotNull(resolveColumn(tr, "c"))))
   }
 
+  test("propagating constraints in aliases") {
+    val tr = LocalRelation('a.int, 'b.string, 'c.int)
+
+    assert(tr.where('c.attr > 10).select('a.as('x), 'b.as('y)).analyze.constraints.isEmpty)
+
+    val aliasedRelation = tr.where('a.attr > 10).select('a.as('x), 'b.as('y), 'a.as('z))
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala]: Should it also produce `a==x`, `b==y` and `a==z`?

Comment (sameeragarwal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala]: good point, these should be a null safe (e.g., a <=> b) right? Added (with the modified test).

**Comment by sameeragarwal** [comment]
test this please

**Comment by marmbrus** [comment]
Merging to master.

**Pull request closed without merging**